### PR TITLE
DispatchKeyDown now uses body documentElement or document for events

### LINKED ISSUE #6952: quick searches no longer works with gestures
**Home Assistant Android app version(s):2026.5.5-full**

**Android version(s):16**

**Device model(s):Google Pixel 10**

**Home Assistant version:2026.6.0b1**

**Last working Home Assistant release (if known):2026.5.4**

**Description of problem, include YAML if issue is related to notifications:**

**Companion App Logs:**

```
--------- beginning of main
06-02 19:42:56.767 25829 25829 D SensorReceiverBase: Received intent: android.app.action.NEXT_ALARM_CLOCK_CHANGED
06-02 19:42:56.777 25829 18705 D SensorReceiverBase$onReceive: Sensor(s) [next_alarm] corresponding to received event android.app.action.NEXT_ALARM_CLOCK_CHANGED are disabled, skipping sensors update
06-02 19:42:56.989 25829 25829 D SensorReceiverBase: Received intent: android.app.action.NEXT_ALARM_CLOCK_CHANGED
06-02 19:42:56.997 25829 19419 D SensorReceiverBase$onReceive: Sensor(s) [next_alarm] corresponding to received event android.app.action.NEXT_ALARM_CLOCK_CHANGED are disabled, skipping sensors update
06-02 19:42:57.254 25829 25829 D SensorReceiverBase: Received intent: android.app.action.NEXT_ALARM_CLOCK_CHANGED
06-02 19:42:57.262 25829 18705 D SensorReceiverBase$onReceive: Sensor(s) [next_alarm] corresponding to received event android.app.action.NEXT_ALARM_CLOCK_CHANGED are disabled, skipping sensors update
06-02 19:42:59.037 25829 18705 D ActivitySensorManager: Received activity update.
06-02 19:42:59.921 25829 20161 D vulkan  : searching for layers in '/data/app/~~WMKbDFy7rSVcj452RhfTsg==/io.homeassistant.companion.android-fpPPXcMeDCE8lq2ofcvnOQ==/lib/arm64'
06-02 19:42:59.922 25829 20161 D vulkan  : searching for layers in '/data/app/~~WMKbDFy7rSVcj452RhfTsg==/io.homeassistant.companion.android-fpPPXcMeDCE8lq2ofcvnOQ==/base.apk!/lib/arm64-v8a'
06-02 19:42:59.922 25829 20161 D vulkan  : searching for layers in '/data/app/~~WMKbDFy7rSVcj452RhfTsg==/io.homeassistant.companion.android-fpPPXcMeDCE8lq2ofcvnOQ==/split_config.arm64_v8a.apk!/lib/arm64-v8a'
06-02 19:42:59.923 25829 20161 D vulkan  : searching for layers in '/data/app/~~WMKbDFy7rSVcj452RhfTsg==/io.homeassistant.companion.android-fpPPXcMeDCE8lq2ofcvnOQ==/split_config.xxhdpi.apk!/lib/arm64-v8a'
06-02 19:43:00.033 25829 25829 D CompatChangeReporter: Compat change id reported: 377864165; UID 10541; state: ENABLED
06-02 19:43:00.060 25829 25829 I ContentCaptureHelper: Setting logging level to OFF
06-02 19:43:00.083 25829 25829 D WindowOnBackDispatcher: setTopOnBackInvokedCallback (unwrapped): android.app.Activity$$ExternalSyntheticLambda0@1afe4bc
06-02 19:43:00.168 25829 25829 D VRI[LaunchActivity]: WindowInsets changed: 1080x2424 statusBars:[0,173,0,0] navigationBars:[0,0,0,63] mandatorySystemGestures:[0,205,0,84] 
06-02 19:43:00.202 25829 25829 D LaunchViewModel: Server (id=3) is connected and registered checking network status
06-02 19:43:00.209 25829 25829 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.211 25829 25829 I LaunchViewModel: Current network state READY_NET_VALIDATED
06-02 19:43:00.217 25829 25851 W JobInfo : Requested important-while-foreground flag for job208327 is ignored and takes no effect
06-02 19:43:00.217 25829 25851 D WM-SystemJobScheduler: Scheduling work ID 290cc71a-7101-4867-8e3f-319691c0a990Job ID 208327
06-02 19:43:00.222 25829 25851 D WM-GreedyScheduler: Starting tracking for 290cc71a-7101-4867-8e3f-319691c0a990
06-02 19:43:00.223 25829 25851 D WM-WorkConstraintsTrack: NetworkRequestConstraintController register shared callback
06-02 19:43:00.225 25829 25901 D WM-WorkConstraintsTrack: NetworkRequestConstraintController onCapabilitiesChanged callback
06-02 19:43:00.225 25829 25901 D WM-WorkConstraintsTrack: Not dispatching constraint state yet: isBlocked=null, capabilitiesInitialized=true
06-02 19:43:00.225 25829 25901 D WM-WorkConstraintsTrack: NetworkRequestConstraintController onBlockedStatusChanged callback false
06-02 19:43:00.225 25829 25856 D WM-GreedyScheduler: Constraints met: Scheduling work ID WorkGenerationalId(workSpecId=290cc71a-7101-4867-8e3f-319691c0a990, generation=0)
06-02 19:43:00.226 25829 25856 D WM-Processor: Processor: processing WorkGenerationalId(workSpecId=290cc71a-7101-4867-8e3f-319691c0a990, generation=0)
06-02 19:43:00.263 25829 25829 D WM-WorkerWrapper: Starting work for io.homeassistant.companion.android.common.util.ResyncRegistrationWorker
06-02 19:43:00.272 25829 18703 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.276 25829 25829 D VRI[LaunchActivity]: visibilityChanged oldVisibility=true newVisibility=false
06-02 19:43:00.277 25829 25829 D WM-SystemJobService: onStartJob for WorkGenerationalId(workSpecId=290cc71a-7101-4867-8e3f-319691c0a990, generation=0)
06-02 19:43:00.277 25829 25857 D WM-Processor: Work WorkGenerationalId(workSpecId=290cc71a-7101-4867-8e3f-319691c0a990, generation=0) is already enqueued for processing
06-02 19:43:00.329 25829 19419 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.339 25829 25829 D WebViewActivity: Current webview package com.google.android.webview and version 148.0.7778.178
06-02 19:43:00.340 25829 25829 I ContentCaptureHelper: Setting logging level to OFF
06-02 19:43:00.342 25829 25851 D WM-Processor: Processor cancelling 5539a26a-e8ae-42e4-a076-2adad895be6b
06-02 19:43:00.342 25829 25851 D WM-Processor: WorkerWrapper could not be found for 5539a26a-e8ae-42e4-a076-2adad895be6b
06-02 19:43:00.345 25829 25851 D WM-GreedyScheduler: Cancelling work ID 5539a26a-e8ae-42e4-a076-2adad895be6b
06-02 19:43:00.345 25829 25829 D WindowOnBackDispatcher: setTopOnBackInvokedCallback (unwrapped): android.app.Activity$$ExternalSyntheticLambda0@d8f7436
06-02 19:43:00.347 25829 25829 D VRI[WebViewActivity]: WindowInsets changed: 1080x2424 statusBars:[0,173,0,0] navigationBars:[0,0,0,63] mandatorySystemGestures:[0,205,0,84] 
06-02 19:43:00.352 25829 25851 W JobInfo : Requested important-while-foreground flag for job208328 is ignored and takes no effect
06-02 19:43:00.352 25829 25851 D WM-SystemJobScheduler: Scheduling work ID f07e24f3-fd57-4227-acaf-99557a135286Job ID 208328
06-02 19:43:00.354 25829 25851 D WM-GreedyScheduler: Starting tracking for f07e24f3-fd57-4227-acaf-99557a135286
06-02 19:43:00.354 25829 25851 D WM-GreedyScheduler: Starting tracking for f07e24f3-fd57-4227-acaf-99557a135286
06-02 19:43:00.355 25829 25851 D WM-WorkConstraintsTrack: NetworkRequestConstraintController send initial capabilities
06-02 19:43:00.355 25829 25851 D WM-GreedyScheduler: Constraints met: Scheduling work ID WorkGenerationalId(workSpecId=f07e24f3-fd57-4227-acaf-99557a135286, generation=0)
06-02 19:43:00.355 25829 25857 D WM-Processor: Processor: processing WorkGenerationalId(workSpecId=f07e24f3-fd57-4227-acaf-99557a135286, generation=0)
06-02 19:43:00.395 25829 19419 D WebSocketCoreImpl: Sending message 709: {type=auth/current_user, id=709}
06-02 19:43:00.419 25829 25829 I ContentCaptureHelper: Setting logging level to OFF
06-02 19:43:00.425 25829 25829 D WebViewActivity$onCreate: Matter/Thread step changed to NOT_STARTED
06-02 19:43:00.426 25829 25902 D WebSocketCoreImpl: Websocket: onMessage (text: HIDDEN)
06-02 19:43:00.426 25829 25902 D WebSocketCoreImpl: Message id 709 received
06-02 19:43:00.430 25829 25829 D WM-SystemJobService: onStartJob for WorkGenerationalId(workSpecId=f07e24f3-fd57-4227-acaf-99557a135286, generation=0)
06-02 19:43:00.431 25829 25829 D WM-WorkerWrapper: Starting work for io.homeassistant.companion.android.sensors.SensorWorker
06-02 19:43:00.431 25829 25855 D WM-Processor: Work WorkGenerationalId(workSpecId=f07e24f3-fd57-4227-acaf-99557a135286, generation=0) is already enqueued for processing
06-02 19:43:00.432 25829 25855 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=290cc71a-7101-4867-8e3f-319691c0a990, tags={ io.homeassistant.companion.android.common.util.ResyncRegistrationWorker } ]
06-02 19:43:00.436 25829 25855 I WM-Processor: Moving WorkSpec (f07e24f3-fd57-4227-acaf-99557a135286) to the foreground
06-02 19:43:00.438 25829 16691 D SensorWorkerBase$doWork: Updating all Sensors in foreground.
06-02 19:43:00.447 25829 25829 D WM-Processor: Processor 290cc71a-7101-4867-8e3f-319691c0a990 executed; reschedule = false
06-02 19:43:00.448 25829 25829 D WM-GreedyScheduler: Stopping tracking for WorkGenerationalId(workSpecId=290cc71a-7101-4867-8e3f-319691c0a990, generation=0)
06-02 19:43:00.448 25829 25829 D WM-SystemJobService: 290cc71a-7101-4867-8e3f-319691c0a990 executed on JobScheduler
06-02 19:43:00.448 25829 25829 I WM-SystemFgDispatcher: Started foreground service Intent { act=ACTION_START_FOREGROUND xflg=0x4 cmp=io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService (has extras) }
06-02 19:43:00.448 25829 25829 D WM-SystemFgDispatcher: Notifying with (id:42, workSpecId: f07e24f3-fd57-4227-acaf-99557a135286, notificationType :1)
06-02 19:43:00.453 25829 25829 D WM-SystemFgDispatcher: Notifying with (id:42, workSpecId: f07e24f3-fd57-4227-acaf-99557a135286, notificationType :1)
06-02 19:43:00.457 25829 25855 D WM-GreedyScheduler: Cancelling work ID 290cc71a-7101-4867-8e3f-319691c0a990
06-02 19:43:00.459 25829 25855 D WM-WorkConstraintsTrack: NetworkRequestConstraintController send initial capabilities
06-02 19:43:00.468 25829 25829 D WindowOnBackDispatcher: setTopOnBackInvokedCallback (unwrapped): null
06-02 19:43:00.469 25829 25829 D ViewRootImpl: Skipping stats log for color mode
06-02 19:43:00.474 25829 25829 D WM-SystemJobService: onStopJob for WorkGenerationalId(workSpecId=290cc71a-7101-4867-8e3f-319691c0a990, generation=0)
06-02 19:43:00.475 25829 25913 D AudioSystem: onNewService: media.audio_policy service obtained 0xb400006f7f656290
06-02 19:43:00.477 25829 25913 D AudioSystem: getService: IAudioPolicyService retrieved: 0xb400006f7f656290  cached: 0xb400006f7f656290
06-02 19:43:00.481 25829 25829 W ServerManagerImpl: Activate with SERVER_ID_ACTIVE is not doing anything
--------- beginning of system
06-02 19:43:00.487 25829 25829 D InsetsController: hide(ime())
06-02 19:43:00.487 25829 25829 I ImeTracker: io.homeassistant.companion.android:8f3a1150: onCancelled at PHASE_CLIENT_ALREADY_HIDDEN
06-02 19:43:00.493 25829 25829 D IntegrationRepositoryImpl: isAppLocked(): false. (LockEnabled: false, appActive: false, expireMillis: 1780348156292, currentMillis: 1780422180492)
06-02 19:43:00.493 25829 25829 D IntegrationRepositoryImpl: isAppLocked(): false. (LockEnabled: false, appActive: false, expireMillis: 1780348156292, currentMillis: 1780422180493)
06-02 19:43:00.493 25829 25829 D IntegrationRepositoryImpl: isAppLocked(): false. (LockEnabled: false, appActive: false, expireMillis: 1780348156292, currentMillis: 1780422180493)
06-02 19:43:00.504 25829 25829 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.504 25829 25829 D ServerConnectionStateProviderImplKt: Using cloud / remote UI URL
06-02 19:43:00.506 25829 25829 D WebViewActivity: Loading HIDDEN (keepHistory true, openInApp true, serverHandleInsets false)
06-02 19:43:00.506 25829 25829 D WindowOnBackDispatcher: setTopOnBackInvokedCallback (unwrapped): androidx.navigationevent.OnBackInvokedInput$createOnBackAnimationCallback$1@c6edb0c
06-02 19:43:00.506 25829 25829 D WindowOnBackDispatcher: setTopOnBackInvokedCallback (unwrapped): android.app.Activity$$ExternalSyntheticLambda0@d8f7436
06-02 19:43:00.516 25829 25829 D WebViewActivity: Waiting for loadedUrl HIDDEN
06-02 19:43:00.519 25829 25829 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.519 25829 25829 D ServerConnectionStateProviderImplKt: Using cloud / remote UI URL
06-02 19:43:00.526 25829 25829 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.526 25829 25829 D ServerConnectionStateProviderImplKt: Using cloud / remote UI URL
06-02 19:43:00.554 25829 25829 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.554 25829 25829 D ServerConnectionStateProviderImplKt: Using cloud / remote UI URL
06-02 19:43:00.601 25829 25920 W VideoCapabilities: Unsupported profile 4 for video/avc
06-02 19:43:00.602 25829 25920 W VideoCapabilities: Unsupported profile 4 for video/avc
06-02 19:43:00.619 25829 25920 W VideoCapabilities: Unsupported profile 4 for video/avc
06-02 19:43:00.619 25829 25920 W VideoCapabilities: Unsupported profile 4 for video/avc
06-02 19:43:00.912 25829 25829 D WebViewActivity: External bus id=1 type=config/get raw=HIDDEN
06-02 19:43:00.914 25829 25829 D WebViewActivity: Sending: externalBus({"id":1,"type":"result","success":true,"result":{"hasSettingsScreen":true,"canWriteTag":true,"hasExoPlayer":true,"canCommissionMatter":true,"canImportThreadCredentials":true,"hasAssist":true,"hasBarCodeScanner":1,"canSetupImprov":true,"downloadFileSupported":true,"appVersion":"2026.5.5-full (22416)","hasEntityAddTo":true,"hasAssistSettings":true}});
06-02 19:43:00.934 25829 25829 D ExternalConfigResponse: Callback from external config (id=1): null
06-02 19:43:00.938 25829 25829 E WebViewActivity: Cannot set status bar color. Skipping coloring...
06-02 19:43:00.938 25829 25829 E WebViewActivity: Cannot set background color. Skipping coloring...
06-02 19:43:00.948 25829 25829 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:00.948 25829 25829 D ServerConnectionStateProviderImplKt: Using cloud / remote UI URL
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: Issue requesting updates for Geolokalisierungssensoren
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: java.lang.Exception: jweu: UNAVAILABLE
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: 	at io.homeassistant.companion.android.sensors.GeocodeSensorManager$getFromLocationAwait$2$1.onError(GeocodeSensorManager.kt:183)
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: 	at android.location.Geocoder$GeocodeCallbackImpl.onError(Geocoder.java:359)
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: 	at android.location.provider.IGeocodeCallback$Stub.onTransact(IGeocodeCallback.java:99)
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: 	at android.os.Binder.execTransactInternal(Binder.java:1471)
06-02 19:43:01.163 25829 17634 E SensorReceiverBase: 	at android.os.Binder.execTransact(Binder.java:1411)
06-02 19:43:01.246 25829 18704 D LocationSensorManager: High accuracy mode disabled, because not in zone [3_zone.oelde_expanded]
06-02 19:43:01.300 25829 25829 D WebViewActivity: External bus id=2 type=connection-status raw=HIDDEN
06-02 19:43:01.335 25829 19419 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:01.936 25829 18705 D ServerConnectionStateProviderImpl: usesInternalSsid is: false, usesWifi is: true
06-02 19:43:01.991 25829 18705 I SensorReceiverBase: Sensor updates and sync completed
06-02 19:43:02.006 25829 25855 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=f07e24f3-fd57-4227-acaf-99557a135286, tags={ io.homeassistant.companion.android.sensors.SensorWorker } ]
06-02 19:43:02.010 25829 25829 D WM-Processor: Processor f07e24f3-fd57-4227-acaf-99557a135286 executed; reschedule = false
06-02 19:43:02.010 25829 25829 D WM-GreedyScheduler: Stopping tracking for WorkGenerationalId(workSpecId=f07e24f3-fd57-4227-acaf-99557a135286, generation=0)
06-02 19:43:02.010 25829 25829 D WM-SystemFgDispatcher: Removing Notification (id: 42, workSpecId: WorkGenerationalId(workSpecId=f07e24f3-fd57-4227-acaf-99557a135286, generation=0), notificationType: 1
06-02 19:43:02.010 25829 25829 D WM-SystemJobService: f07e24f3-fd57-4227-acaf-99557a135286 executed on JobScheduler
06-02 19:43:02.019 25829 25855 D WM-GreedyScheduler: Cancelling work ID f07e24f3-fd57-4227-acaf-99557a135286
06-02 19:43:02.022 25829 25855 D WM-SystemJobScheduler: Scheduling work ID f07e24f3-fd57-4227-acaf-99557a135286Job ID 208329
06-02 19:43:02.024 25829 25855 D WM-WorkConstraintsTrack: NetworkRequestConstraintController unregister shared callback
06-02 19:43:03.218 25829 25829 D WebViewActivity: External bus id=3 type=theme-update raw=HIDDEN
06-02 19:43:03.333 25829 25829 D WebViewActivity: External bus id=4 type=theme-update raw=HIDDEN
06-02 19:43:03.440 25829 25829 D WebViewActivity$getAndSetStatusBarNavigationBarColors: Color from webview is "#101419-SPACER-#1d2128"
06-02 19:43:03.440 25829 25829 D WebViewActivity$getAndSetStatusBarNavigationBarColors: Color from webview is "#101419-SPACER-#1d2128"
06-02 19:43:03.440 25829 19418 D WebViewPresenterImpl$parseWebViewColor: Try getting color from webview color "#101419".
06-02 19:43:03.440 25829 19418 I WebViewPresenterImpl$parseWebViewColor: Found color -15723495.
06-02 19:43:03.441 25829 17634 D WebViewPresenterImpl$parseWebViewColor: Try getting color from webview color "#101419".
06-02 19:43:03.441 25829 17634 I WebViewPresenterImpl$parseWebViewColor: Found color -15723495.
06-02 19:43:03.441 25829 17634 D WebViewPresenterImpl$parseWebViewColor: Try getting color from webview color "#1d2128".
06-02 19:43:03.441 25829 17634 I WebViewPresenterImpl$parseWebViewColor: Found color -14868184.
06-02 19:43:03.441 25829 17634 D WebViewPresenterImpl$parseWebViewColor: Try getting color from webview color "#1d2128".
06-02 19:43:03.441 25829 17634 I WebViewPresenterImpl$parseWebViewColor: Found color -14868184.
06-02 19:43:03.534 25829 25829 E WebViewActivity$onCreate: onReceivedHttpError: 404 :  for: https://cdn.jsdelivr.net/gh/nerwyn/material-rounded-theme@main/dist/material-rounded-theme.js
06-02 19:43:03.534 25829 25829 D WebViewActivity: External bus id=5 type=theme-update raw=HIDDEN
06-02 19:43:03.708 25829 25829 D WebViewActivity$getAndSetStatusBarNavigationBarColors: Color from webview is "#101419-SPACER-#1d2128"
06-02 19:43:03.709 25829 18705 D WebViewPresenterImpl$parseWebViewColor: Try getting color from webview color "#101419".
06-02 19:43:03.709 25829 18705 I WebViewPresenterImpl$parseWebViewColor: Found color -15723495.
06-02 19:43:03.709 25829 18704 D WebViewPresenterImpl$parseWebViewColor: Try getting color from webview color "#1d2128".
06-02 19:43:03.709 25829 18704 I WebViewPresenterImpl$parseWebViewColor: Found color -14868184.
06-02 19:43:03.836 25829 25913 D AudioSystem: onNewServiceWithAdapter: media.audio_flinger service obtained 0xb400006f8f6f3f40
06-02 19:43:03.847 25829 25913 D AudioSystem: getService: IAudioFlingerService retrieved: 0xb400006f7f63fa40  IAudioFlinger cached: 0xb400006f8f6f3f40
06-02 19:43:04.600 25829 25913 I AAudio  : AAudioStreamBuilder_openStream() called ----------------------------------------
06-02 19:43:04.602 25829 25913 I AudioStreamBuilder: rate   =  44100, channels  = 2, channelMask = 0x3, format   = 0x5, sharing = SH, dir = OUTPUT
06-02 19:43:04.602 25829 25913 I AudioStreamBuilder: devices = AUDIO_PORT_HANDLE_NONE, sessionId = -1, perfMode = 11, callback: ON with frames = 1024
06-02 19:43:04.602 25829 25913 I AudioStreamBuilder: usage  =      1, contentType = 0, inputPreset = 0, allowedCapturePolicy = 0
06-02 19:43:04.602 25829 25913 I AudioStreamBuilder: privacy sensitive = false, opPackageName = (null), attributionTag = (null)
06-02 19:43:04.602 25829 25913 D AudioStreamBuilder: build, global mmap policy is 0
06-02 19:43:04.611 25829 25913 D AudioStreamBuilder: build, system mmap policy is 2
06-02 19:43:04.611 25829 25913 D AudioStreamBuilder: build, final mmap policy is 2
06-02 19:43:04.612 25829 25913 D AudioStreamBuilder: build, system mmap exclusive policy is 2
06-02 19:43:04.612 25829 25913 D AudioStreamBuilder: build, final mmap exclusive policy is 2
06-02 19:43:04.612 25829 25913 D AudioStreamBuilder: build() MMAP not used because AAUDIO_PERFORMANCE_MODE_LOW_LATENCY or AAUDIO_PERFORMANCE_MODE_POWER_SAVING_OFFLOADED not requested.
06-02 19:43:04.613 25829 25913 D mpanion.android: PlayerBase::PlayerBase()
06-02 19:43:04.614 25829 25913 D AudioStreamTrack: open(), request notificationFrames = 0, frameCount = 0
06-02 19:43:04.647 25829 25913 D AAudioStream: setState(s#1) from 0 to 2
06-02 19:43:04.649 25829 25913 I AAudio  : AAudioStreamBuilder_openStream() got Legacy, devIds = [3], perf = PS, burst = 1768
06-02 19:43:04.649 25829 25913 I AAudio  : AAudioStreamBuilder_openStream() returns 0 = AAUDIO_OK for s#1
```

**Screenshot or video of problem:**

**Additional information:**

### PR DESCRIPTION
**Pull request by timoschwarzer**
## Summary
This PR makes synthetic keyboard events be dispatched on `document.body` instead of `document` due to a bug in the `tinykeys` library used in the HA frontend. This fixes search gestures (any everything else that would dispatch keyboard events).

Fixes https://github.com/home-assistant/android/issues/6952

## Checklist

- [ ] New or updated tests have been added to cover the changes following the testing [guidelines](https://developers.home-assistant.io/docs/android/testing/introduction).
- [x] The code follows the project's [code style](https://developers.home-assistant.io/docs/android/codestyle) and [best_practices](https://developers.home-assistant.io/docs/android/best_practices).
- [ ] The changes have been thoroughly tested, and edge cases have been considered.
  - I did *not* compile and test the app due to lack of a development environment.
  - Since this change only changes JavaScript that is run in the HA frontend, I have run these in a browser console in my HA instance:
    - `document.dispatchEvent(new KeyboardEvent('keydown', {key: 'd', code: 'KeyD', keyCode: 68, which: 68, ctrlKey: false, bubbles: true, cancelable: true}))` → crashes
    - `document.body.dispatchEvent(new KeyboardEvent('keydown', {key: 'd', code: 'KeyD', keyCode: 68, which: 68, ctrlKey: false, bubbles: true, cancelable: true}))` → works
- [x] Changes are backward compatible whenever feasible. Any breaking changes are documented in the changelog for users and/or in the code for developers depending on the relevance.

### CODE REVIEW

**Comment by timoschwarzer** [comment]
Hmm, I can't connect the dots between the CI failing and my changes, so I guess the failures are not related 🤔

**Comment by jpelgrom** [comment]
The CI failure is indeed unrelated, you can ignore it for now. (#7088)

**Comment by TimoPtr** [comment]
I've updated your branch it should fix the CI issue

**Comment by TimoPtr** [comment]
Does the change works on old version of the frontend too? I see that you've open an issue on the upstream, could a fix break this change?

**Comment by timoschwarzer** [comment]
My upstream fix should not break this workaround because `document.body` is an `HTMLElement` and the library works for `KeyboardEvent`s that target any `HTMLElement` in the DOM. (`KeyboardEvent`s that are produced by a real keyboard input usually target some focused element inside the DOM and bubble up from there). I cannot say for sure whether old versions of the HA frontend work with this workaround, it should be relatively easy to test though: Run an old HA version and run `document.body.dispatchEvent(new KeyboardEvent('keydown', {key: 'd', code: 'KeyD', keyCode: 68, which: 68, ctrlKey: false, bubbles: true, cancelable: true}))` in the browser console. If the device search shows up, it works. I didn't trace down the exact version of the HA frontend that first included the incompatible version of `tinykeys`, I have only tested with the current HA 2026.6.4 (Frontend 20260527.7).

File: app/src/main/kotlin/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 1]
```diff
@@ -2219,7 +2219,7 @@ class WebViewActivity :
             bubbles: true,
             cancelable: true
         });
-        document.dispatchEvent(event);
+        document.body.dispatchEvent(event);
```
Comment (TimoPtr) [inline app/src/main/kotlin/io/homeassistant/companion/android/webview/WebViewActivity.kt]:
```suggestion
         (document.body || document.documentElement || document).dispatchEvent(event);
```

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/gesture/FrontendGestureManager.kt [hunk 2]
```diff
@@ -124,7 +124,7 @@ class FrontendGestureManager @Inject constructor(
                 bubbles: true,
                 cancelable: true
             });
-            document.dispatchEvent(event);
+            document.body.dispatchEvent(event);
```
Comment (TimoPtr) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/gesture/FrontendGestureManager.kt]:
```suggestion
             (document.body || document.documentElement || document).dispatchEvent(event);
```

**Comment by aalaei** [comment]
Thanks for this fix @timoschwarzer ! I've been stuck on the stable Play Store release (2026.6.5) where all search-related gestures (quick search, search entities, search devices, search commands) silently fail while other gesture actions like server list work fine. This PR is exactly the fix I need.

It's been over two months since the last stable release was promoted to the Play Store. Is there any timeline for when a new stable version will land? I'd prefer not to sideload a pre-release.

**Comment by jpelgrom** [comment]
> It's been over two months since the last stable release was promoted to the Play Store. Is there any timeline for when a new stable version will land? 

The plan is to release a stable version in 1-2 weeks. Google hasn't approved any of our latest 3 betas yet though so it might take longer.
